src/StaticWebAssetsSdk/Tasks/DefineStaticWebAssetEndpoints.cs

```diff
@@ -43,41 +43,23 @@ public override bool Execute()
         var contentTypeMappings = ContentTypeMappings.Select(ContentTypeMapping.FromTaskItem).OrderByDescending(m => m.Priority).ToArray();
         var contentTypeProvider = new ContentTypeProvider(contentTypeMappings);
-        var endpoints = new ConcurrentBag<StaticWebAssetEndpoint>();
-
-        Parallel.For(0, CandidateAssets.Length, i =>
-        {
-            var asset = StaticWebAsset.FromTaskItem(CandidateAssets[i]);
-            var routes = asset.ComputeRoutes().ToList();
-
-            if (existingEndpointsByAssetFile != null && existingEndpointsByAssetFile.TryGetValue(asset.Identity, out var set))
-            {
-                for (var j = routes.Count - 1; j >= 0; j--)
-                {
-                    var (label, route, values) = routes[j];
-                    // StaticWebAssets has this behavior where the base path for an asset only gets applied if the asset comes from a
-                    // package or a referenced project and ignored if it comes from the current project.
-                    // When we define the endpoint, we apply the path to the asset as if it was coming from the current project.
-                    // If the endpoint is then passed to a referencing project or packaged into a nuget package, the path will be
-                    // adjusted at that time.
-                    var finalRoute = asset.IsProject() || asset.IsPackage() ? StaticWebAsset.Normalize(Path.Combine(asset.BasePath, route)) : route;
-
-                    // Check if the endpoint we are about to define already exists. This can happen during publish as assets defined
-                    // during the build will have already defined endpoints and we only want to add new ones.
-                    if (set.Contains(finalRoute))
-                    {
-                        Log.LogMessage(MessageImportance.Low, $"Skipping asset {asset.Identity} because an endpoint for it already exists at {route}.");
-                        routes.RemoveAt(j);
-                    }
-                }
-            }
-
-            foreach (var endpoint in CreateEndpoints(routes, asset, contentTypeProvider))
-            {
-                Log.LogMessage(MessageImportance.Low, $"Adding endpoint {endpoint.Route} for asset {asset.Identity}.");
-                endpoints.Add(endpoint);
-            }
-        });
+        var endpoints = new List<StaticWebAssetEndpoint>();
+
+        Parallel.For(
+            0,
+            CandidateAssets.Length,
+            () => new ParallelWorker(
+                endpoints,
+                new List<StaticWebAssetEndpoint>(),
+                CandidateAssets,
+                existingEndpointsByAssetFile,
+                Log,
+                contentTypeProvider,
+                _assetFileDetails,
+                TestLengthResolver,
+                TestLastWriteResolver),
+            static (i, loop, state) => state.Process(i, loop),
+            static worker => worker.Finally());
```

Comment:
This is cleaning up the implementation to avoid excessive locking by storing the intermediate results into a local list for each worker and saving them to the final collection at the end of processing.